---
layout:     post
title:      HADDOCK2.4安装及使用
subtitle:   HADDOCK
date:       2021-09-25
author:     xpgege
header-img: img/post-web.jpg
catalog: true
tags:
	- 蛋白质
---



# 1.HADDOCK

HADDOCK（High  Ambiguous Driven biomolecular Docking ） [官网](https://www.bonvinlab.org/education/)

是从 ARIA (https://aria.pasteur.fr) 派生的 Python 脚本集合，这些脚本利用了 CNS（晶体学和核磁共振系统）的力量– https://cns-online.org) 用于分子复合物的结构计算。 HADDOCK 与其他对接软件的不同之处在于它从 CNS 继承而来的能力，将实验数据作为约束并使用这些数据来指导对接过程以及传统的能量学和形状互补。此外，与 CNS 的密切耦合使 HADDOCK 能够实际生成足够质量的模型，以便在蛋白质数据库中存档。

## Before docking

### [How to prepare structures for HADDOCK?](https://www.bonvinlab.org/software/bpg/structures/)



[![img](https://www.bonvinlab.org/software/bpg/pdb.png)](https://www.bonvinlab.org/software/bpg/structures)



### [How to use information about interactions in HADDOCK?](https://www.bonvinlab.org/software/bpg/restraints/)



[![img](https://www.bonvinlab.org/software/bpg/restraints.png)](https://www.bonvinlab.org/software/bpg/restraints)



------

## Docking with HADDOCK

| **[Small molecules](https://www.bonvinlab.org/software/bpg/small_mol)** | **[Glycans](https://www.bonvinlab.org/software/bpg/glycans)** | **[Peptide](https://www.bonvinlab.org/software/bpg/peptides)** | **[Nucleotides](https://www.bonvinlab.org/software/bpg/dna)** | **[Proteins](https://www.bonvinlab.org/software/bpg/proteins)** | **[Coarse grain](https://www.bonvinlab.org/software/bpg/cg)** |
| ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
| [![img](https://www.bonvinlab.org/software/bpg/small_mol.png)](https://www.bonvinlab.org/software/bpg/small_mol) | [![img](https://www.bonvinlab.org/software/bpg/sugar.png)](https://www.bonvinlab.org/software/bpg/glycans) | [![img](https://www.bonvinlab.org/software/bpg/pep.png)](https://www.bonvinlab.org/software/bpg/peptides) | [![img](https://www.bonvinlab.org/software/bpg/dna.png)](https://www.bonvinlab.org/software/bpg/dna) | [![img](https://www.bonvinlab.org/software/bpg/antigen.png)](https://www.bonvinlab.org/software/bpg/proteins) | [![img](https://www.bonvinlab.org/software/bpg/cg.png)](https://www.bonvinlab.org/software/bpg/cg) |
| ![img](https://www.bonvinlab.org/software/bpg/arrow_rainbow.png) |                                                              |                                                              |                                                              |                                                              |                                                              |

------

## [After docking: How to analyse docking results from HADDOCK or refine models?](https://www.bonvinlab.org/software/bpg/analysis/)



[![img](https://www.bonvinlab.org/software/bpg/analysis.png)](https://www.bonvinlab.org/software/bpg/analysis)



# 2. Install

[Installation Instructions：reference](https://www.bonvinlab.org/software/haddock2.4/installation/)

##   2.1 Downloading HADDOCK

```
https://www.bonvinlab.org/software/haddock2.4/download

需要填写对应的邮箱和单位（我填的是学校的教育邮箱），经过验证后会发一个邮件，密码和下载链接

我这里附在这里：

url：https://surfdrive.surf.nl/files/index.php/s/OQ6jMGdcyW7k49k

password：HADDOCK2.4-May-2021-spring-is-here

```



##   2.2 Downloading CNS

```shell
http://cns-online.org/cns_request/
同样需要填写邮箱和详细信息，获得账号密码下载

This password is changed every Saturday at 6:15 EST.

To download the CNS distribution from a web browser:

        http://cns-online.org/download/ 
        login: download
        password: u2bepah

下载 cns_solve_1.3_all_intel-mac_linux.tar.gz
tar -zxvf cns_solve_1.3_all_intel-mac_linux.tar.gz
cd cns_solve_1.3
#这里需要csh或zcsh 的shell （sudo apt-get install csh）
vim cns_solve_env （将这一行改成 setenv CNS_SOLVE '~/software/cns_solve_1.3'）
make install

#before using CNSsolve:
source cns_solve_env


#重新compile CNS
Marija
Oct '20
Thank you very much for help, everything works now.
Below I just summarize what I have done to install CNS v1.3 on Centos7.7 and to link CNS to the local version of HADDOCK2.4:
-> After registration at http://cns-online.org/cns_request/ 5, download cns_solve_xyz.tar.gz file “xyz = cns_solve downloaded version”
-> Enter the command in terminal: gunzip cns_solve_xyz.tar.gz
-> Then enter: tar xvf cns_solve_xyz.tar
-> This will create a new cns_solve_13 directory. Now copy all files from haddock2.4/cns1.3/ to cns_solve_1.3/source
-> Now enter into cns_solve_1.3 directory and type: make install
-> Copy the path of the directory.
-> Now open the file cns_solve_env and paste the copied path in place of ‘location of CNSsolve directory’
-> Go to the directory cns_solve_1.3/intel-x86_64bit-linux/source
-> Type: make cns_solve
-> Then go back to cns_solve_1.3 directory, and enter the command: tcsh
-> Then you entered into C-shell, now enter the command: source cns_solve_env
-> Then enter: cns
-> Your software should run without any error.

cp ../haddock2.4/cns1.3/* ./source/
make install
vim .cns_solve_env_sh
source .cns_solve_env_sh
```



##   2.3 Recommended auxiliary software

#### FreeSASA

```
mkdir ~/software
cd ~/software
wget https://freesasa.github.io/freesasa-2.0.3.tar.gz
tar xvfz freesasa-2.0.3.tar.gz
cd freesasa-2.0.3
./configure \-\-disable-json \-\-disable-xml \-\-prefix ~/software
make
make install
```

#### HADDOCK-tools

```
  cd ~/software
  git clone https://github.com/haddocking/haddock-tools
```

#### MolProbity

```
https://github.com/rlabduke/MolProbity/blob/master/bin/linux/reduce
直接从链接下载reduce可执行文件
然后放置到~/software/bin里
chmod +x ~/software/bin/reduce
```

#### Pymol


```
#如果是ubantu 可以直接
sudo apt-get install pymol
#服务器系统是centos
参考:https://zhuanlan.zhihu.com/p/88113746
```

#### PDB-tools

```
cd  ~/software
git clone https://github.com/haddocking/pdb-tools
```

#### ProFit

```
详情见INSTALL文件

tar -zxvf profit.tar.gz
cd  ProFit_V3.3/src/
make
```



##   2.4 Configuring HADDOCK

````
tar -zxvf haddock2.4-2021-05.tgz
cd haddock2.4-2021-05
vim config.local

```
set CNSTMP=/home/xpgege/software/cns_solve_1.3/intel-x86_64bit-linux
set QUEUETMP=/bin/csh
set NUMJOB=4
set QUEUESUB=QueueSubmit_concat.py
```
根据自己情况修改参数

./install.csh config.local

vim Haddock/Main/MHaddock.py 

```
jobconcat["0"] =  1
jobconcat["1"] = 1
jobconcat["2"] = 1
If running on local system, change all values to 1.
```
 
cd ~/software/haddock2.4-2021-05
source ./haddock_configure.sh
````





# 3. prepare structures for HADDOCK

##   3.1 Experimental  Structure

- Worldwide Protein Data Bank [wwPDB](http://www.wwpdb.org/)
- Protein Data Bank in Europe [PDBe](https://www.ebi.ac.uk/pdbe/node/1)
- The Research Collaboratory for Structural Bioinformatics Protein Data Bank [RCSB PDB](https://www.rcsb.org/)
- Protein Data Bank Japan [PDBj](https://pdbj.org/)
- Biological Magnetic Resonance Data Bank [BMRB](http://www.bmrb.wisc.edu/)



##   3.2  Sequence and homologous proteins

**homologue search** are here:

- [UniPROT](https://www.uniprot.org/blast/)
- [HMMER](https://www.ebi.ac.uk/Tools/hmmer/search/phmmer)

**homology model building** are here:

- SWISS-MODEL

  - this online tool can both look for homologous proteins and build a protein model

- MODELLER

  - online version [ModLoop](https://modbase.compbio.ucsf.edu/modloop/) for loop modeling
  - local version for homology or comparative modeling of protein three-dimensional structure

  

  

##   3.3  Modeling of small molecules

 

- OpenEye OMEGA

  - OMEGA uses the SMILES strings as input to generate three-dimensional (3D) conformations of ligands. 

    OMEGA was used by our group in previous rounds of the D3R challenge.

- RDKit

  - open source chemoinformatics and machine learning software

![img](https://www.bonvinlab.org/software/bpg/ligand.png)

- to prepare topology and parameter files for the ligand in CNS format one can use:
  - the **PRODGR** server maintained by Daan van Aalten at Dundee University: [https://prodrg2.dyndns.org](https://prodrg2.dyndns.org/)
    This server allows you to draw your molecule or paste coordinates and will return topologies and parameter files in various format, including CNS. You should turn on the electrostatic to obtain partial charges.
  - the **Automated Topology Builder (ATB)** and Repository developed in the group of Prof. Alan Mark at the University of Queensland in Brisbane: https://compbio.biosci.uq.edu.au/atb

Preparation of small molecules for docking is further described in the [frequently asked questions page](https://www.bonvinlab.org/software/haddock2.4/faq/#small-ligand-docking-with-haddock).



##   3.3  Molecular Dynamics

​	

- [GROMACS](http://www.gromacs.org/)
- [GROMOS](http://www.gromos.net/)
- [AMBER](https://ambermd.org/)
- [CHARMM](https://www.charmm.org/)
- [OpenMM](http://openmm.org/)

Examples of using MD for HADDOCK are shown here:

- [Molecular Dynamics Simulation of the p53 N-terminal peptide](https://www.bonvinlab.org/education/molmod/simulation/)
- [Tutorial EDES / HADDOCK for ligand-protein docking](https://www.bonvinlab.org/education/biomolecular-simulations-2019/Metadynamics_tutorial/)



##   3.4  Editing pdb files

- [PDB-Tools Web](https://bianca.science.uu.nl/pdbtools/)
- [Local version of PDB-Tools](http://www.bonvinlab.org/pdb-tools/)

Tutorials:

- [Tutorial explaining use of PDB-Tools Web](https://www.bonvinlab.org/education/HADDOCK24/HADDOCK24-antibody-antigen/#inspecting-and-preparing-the-antibody-for-docking)
- [Tutorial explaining use of the local version of PDB-Tools](https://www.bonvinlab.org/education/HADDOCK24/HADDOCK24-local-tutorial/#preparing-pdb-files-for-docking)





# 4.  Tutorial

## 4.1 Cleaning PDB files prior to docking

```
#打开pymol
fetch 1F3G
show cartoon
hide lines
show sticks, resn HIS #显示组氨酸结构

zoom resn HIS 
zoom vis
remove resn HOH
save e2a_1F3G.pdb

```

```
#shell
pdb_mutate.py e2a_1F3G.pdb A 90 HIS NEP >e2aP_1F3G.pdb
pdb_chain.py e2aP_1F3G.pdb | ~/software/pdb-tools/pdb_seg.py >e2aP_1F3G-clean.pdb
pdb_validate.py e2aP_1F3G-clean.pdb
# Line 1116 is short: 3 < 80
#1116行为END 应该没问题
pdb_gap.py e2aP_1F3G-clean.pdb

```

## 4.2 Dealing with an ensemble of models

```
fetch 1HDN
show cartoon
set all_states, on
#共有30个ensemble sturcture
save hpr_1HDN.pdb, state=0
```

```
pdb_chain.py hpr_1HDN.pdb | pdb_seg.py >hpr_1HDN-clean.pdb
pdb_splitmodel.py hpr_1HDN-clean.pdb
ls -rt hpr_1HDN-clean_*.pdb | head -10 > hpr_1HDN_1.list
```

## 4.3 Dealing with multi-chain proteins

```
fetch 4G6K

show cartoon
hide lines
remove resn HOH

save 4G6K.pdb
```

```
pdb_selchain.py -H 4G6K.pdb >4G6K_H.pdb

pdb_selchain.py -L 4G6K.pdb | pdb_shiftres.py -500 >4G6K_L.pdb

cat 4G6K_H.pdb 4G6K_L.pdb | pdb_chain.py | pdb_seg.py > 4G6K-clean.pdb

#提取 L H链 合并
#define a few distance restraints to keep them together during the high temperature flexible refinement stage of HADDOC

restrain_bodies.py 4G6K-clean.pdb >antibody-unambig.tbl

  assign (segid  and resi 189 and name CA) (segid  and resi 693 and name CA) 21.023 0.0 0.0
  assign (segid  and resi 116 and name CA) (segid  and resi 702 and name CA) 44.487 0.0 0.0
```



## 4.4  Defining restraints for docking

Distance restraints are defined as:

```
  assi (selection1) (selection2) distance, lower-bound correction, upper-bound correction
```

ambiguous interaction restraints (AIRs)

specific distance restraints



 it is important to define the residues at the interface for each molecule based on NMR chemical shift perturbation（扰动） data, mutagenesis（诱变） data or any kind of data that provides information on the interaction interface.



- The *"active"* residues are those experimentally identified to be involved in the interaction between the two molecules

  AND solvent accessible (either main chain or side chain relative accessibility should be typically > 40-50%).

  

- The ***"passive"\*** residues are all solvent accessible surface neighbors of active residues.



### 4.4.1 Defining AIRs from interface mapping

```
pymol e2a_1F3G.pdb

color white, all
show surface
select e2a_active, (resi 38,40,45,46,69,71,78,80,94,96,141)
color red, e2a_active

select e2a_passive, (resi 35,37,39,42,43,44,47,48,64,66,68,72,73,74,82,83,84,86,97,99,100,105,109,110,112,131,132,133,143,144)
color green, e2a_passive
```

![img](https://www.bonvinlab.org/education/HADDOCK24/HADDOCK24-local-tutorial/e2a-active-passive.png)

```
freesasa e2a_1F3G.pdb --format=rsa >e2a_1F3G.rsa

awk '{if (NF==13 && $5>40) print $0; if (NF==14 && $6>40) print $0}' e2a_1F3G.rsa
```

![image-20210930100710801](https://i.loli.net/2021/09/30/SciN7npGmQzFjfl.png)

To use our `haddock-tools` `active-passive-to-ambig.py` script you need to create for each molecule a file containing two lines:

- The first line corresponds to the list of active residues (numbers separated by spaces)
- The second line corresponds to the list of passive residues.

```
e2a-act-pass.list
38 40 45 46 69 71 78 80 94 96 141
35 37 39 42 43 44 47 48 64 66 68 72 73 74 82 83 84 86 97 99 100 105 109 110 112 131 132 133 143 144

hpr-act-pass.list
15 16 17 20 48 49 51 52 54 56
9 10 11 12 21 24 25 34 37 38 40 41 43 45 46 47 53 55 57 58 59 60 84 85

active-passive-to-ambig.py e2a-act-pass.list hpr-act-pass.list >e2a-hpr-ambig.tbl
```



## 4.5 Haddock

### acitivate environment

```
conda activate py2
source /home/xpgege/software/cns_solve_1.3/.cns_solve_env_sh
source /home/xpgege/software/haddock2.4/haddock_configure.sh
export PATH=$PATH:/home/xpgege/software/ProFit_V3.3/src/
```

### edit run parameter

```
#general format
AMBIG_TBL=antibody-antigen-ambig.tbl
HADDOCK_DIR=PATH/TO/HADDOCK/INSTALLATIONDIR/haddock2.4
N_COMP=2
PDB_FILE1=4G6K-clean.pdb
PDB_FILE2=4I1B.pdb
PROJECT_DIR=./
PROT_SEGID_1=A
PROT_SEGID_2=B
RUN_NUMBER=1
UNAMBIG_TBL=antibody-antigen-unambig.tbl
```

### first run

```
haddock2.4
```

![image-20211109171208438](https://i.loli.net/2021/11/09/WZvKeBx42yDgwXI.png)

### second run

```
cd run1
vim run.cns # change parameter
haddock2.4
```

![image-20211109195905154](https://i.loli.net/2021/11/09/i2rd5LDlaC3PI76.png)

### structure analysis

```
cd structures/it1/water
```

`file.list` which contains the sorted PDB models with their corresponding HADDOCK score, e.g.:

```
"PREVIT:protein-dna_6w.pdb"  { -87.01708 }
"PREVIT:protein-dna_14w.pdb"  { -84.65586 }
"PREVIT:protein-dna_1w.pdb"  { -82.97487 }
"PREVIT:protein-dna_4w.pdb"  { -82.18431 }
"PREVIT:protein-dna_12w.pdb"  { -80.76248 }
"PREVIT:protein-dna_5w.pdb"  { -77.70296 }
"PREVIT:protein-dna_15w.pdb"  { -75.7234 }
"PREVIT:protein-dna_3w.pdb"  { -72.927 }
"PREVIT:protein-dna_20w.pdb"  { -71.47823 }
"PREVIT:protein-dna_17w.pdb"  { -71.17675 }
"PREVIT:protein-dna_9w.pdb"  { -63.33197 }
"PREVIT:protein-dna_10w.pdb"  { -60.4638 }
"PREVIT:protein-dna_11w.pdb"  { -59.15541 }
"PREVIT:protein-dna_2w.pdb"  { -58.38473 }
"PREVIT:protein-dna_13w.pdb"  { -56.21771 }
"PREVIT:protein-dna_8w.pdb"  { -53.54629 }
"PREVIT:protein-dna_7w.pdb"  { -50.4619 }
"PREVIT:protein-dna_16w.pdb"  { -47.37582 }
"PREVIT:protein-dna_19w.pdb"  { -37.65482 }
"PREVIT:protein-dna_18w.pdb"  { -34.17352 }


```

Each PDB model contains a header section with the various energy terms, e.g.:

```
REMARK FILENAME="protein-dna_1w.pdb0"
REMARK ===============================================================
REMARK HADDOCK run for protein-dna
REMARK initial structure: protein-dna_1.pdb
REMARK ===============================================================
REMARK            total,bonds,angles,improper,dihe,vdw,elec,air,cdih,coup,rdcs,vean,dani,xpcs,rg
REMARK energies: -295.594, 0, 0, 0, 0, -36.7761, -301.685, 42.8668, 0, 0, 0, 0, 0, 0, 0
REMARK ===============================================================
REMARK            bonds,angles,impropers,dihe,air,cdih,coup,rdcs,vean,dani,xpcs
REMARK rms-dev.: 0,0,0,0,1.03402,0,0, 0, 0, 0, 0
REMARK ===============================================================
REMARK               air,cdih,coup,rdcs,vean,dani,xpcs
REMARK               >0.3,>5,>1,>0,>5,>0.2,>0.2
REMARK violations.: 1, 0, 0, 0, 0, 0, 0
REMARK ===============================================================
REMARK                        CVpartition#,violations,rms
REMARK AIRs cross-validation: 2, 8, 2.55642
REMARK ===============================================================
REMARK NCS energy: 0
REMARK ===============================================================
REMARK Symmetry energy: 0
REMARK ===============================================================
REMARK Membrane restraining energy: 0
REMARK ===============================================================
REMARK Local cross-correlation:  0.0000
REMARK ===============================================================
REMARK Desolvation energy: 9.85155
REMARK Internal energy free molecules: -3675.58
REMARK Internal energy complex: -3636.42
REMARK Binding energy: -289.452
REMARK ===============================================================
REMARK buried surface area: 1084.23
REMARK ===============================================================
REMARK water - chain_1: 0 0 0
REMARK water - chain_2: 0 0 0
REMARK ===============================================================
REMARK water - water: 0 0 0
REMARK ===============================================================
REMARK DATE:09-Nov-2021  17:29:14       created by user: xpgege
REMARK VERSION:1.3U
ATOM      1  N   MET A  -1     -19.154  -2.189   9.302  1.00 10.00      A    N  
ATOM      2  HN  MET A  -1     -20.021  -1.805   9.571  1.00 10.00      A    H  
ATOM      3  CA  MET A  -1     -17.931  -1.517   9.727  1.00 10.00      A    C  
```

```
$HADDOCKTOOLS/ana_structures.csh

$HADDOCKTOOLS/ana_clusters.csh -best 4 analysis/cluster.out
```

