---
layout:     post
title:      qiime2 16S数据分析
subtitle:   qiime2
date:       2020-6-15
author:     jfh
header-img: img/post-web.jpg
catalog: true
tags:
    - 数据分析
---

# 合并孟加拉和马拉维数据分析

## 合并数据

将孟加拉和马拉维的所有数据进行合并，统一分析。首先创建所有健康样本的`manifest`数据。然后导入到QIIME2中。文件分析路径位于`Storage3/Micro-Develop/ref7_13`下

``` bash
# 激活环境
export LC_ALL=en_US.utf-8
export LANG=en_US.utf-8
conda activate qiime2-2020.2

# 将双端测序数据输入
qiime tools import --type 'SampleData[PairedEndSequencesWithQuality]' \
--input-path Healthy_cohort_manifest.txt --output-path pe_sequence.qza \
--input-format PairedEndFastqManifestPhred33V2

qiime vsearch join-pairs \
  --i-demultiplexed-seqs pe_sequence.qza \
  --o-joined-sequences pe-joined.qza
  
# 获取数据的质量图像
qiime demux summarize --i-data pe-joined.qza \
--o-visualization pe_sequence.qzv
```

### 数据质控

合并后的数据的质量如图所示

![](https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/16S/Micor-develop/ref7_13/combined_reads_quality.png)

```bash
qiime quality-filter q-score-joined \
  --i-demux pe-joined.qza \
  --p-min-quality 20 \
  --o-filtered-sequences demux-joined-filtered.qza \
  --o-filter-stats demux-joined-filter-stats.qza
  
qiime deblur denoise-16S \
  --i-demultiplexed-seqs demux-joined-filtered.qza \
  --p-trim-length 220 \
  --p-sample-stats \
  --o-representative-sequences rep-seqs.qza \
  --o-table table.qza \
  --o-stats deblur-stats.qza
  
qiime feature-table summarize \
  --i-table table.qza \
  --o-visualization table.qzv
```

| Metric             | Sample     |
| :----------------- | :--------- |
| Number of samples  | 1,262      |
| Number of features | 3,546      |
| Total frequency    | 22,182,088 |

### Frequency per sample

|                   |      Frequency      |
| :---------------- | :-----------------: |
| Minimum frequency |        360.0        |
| 1st quartile      |       7,507.0       |
| Median frequency  |      12,222.0       |
| 3rd quartile      |      18,097.0       |
| Maximum frequency |      561,071.0      |
| Mean frequency    | 17,576.931854199684 |

### 稀释 

以3000为标准进行稀释。

![](https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/16S/Micor-develop/ref7_13/%E7%A8%80%E9%87%8A%E6%9B%B2%E7%BA%BF.png)

``` bash
## 进化树
qiime phylogeny align-to-tree-mafft-fasttree \
  --i-sequences rep-seqs.qza \
  --o-alignment aligned-rep-seqs.qza \
  --o-masked-alignment masked-aligned-rep-seqs.qza \
  --o-tree unrooted-tree.qza \
  --o-rooted-tree rooted-tree.qza

qiime diversity alpha-rarefaction \
  --i-table table.qza \
  --i-phylogeny rooted-tree.qza \
  --p-max-depth 3000 \
  --m-metadata-file metadata.txt \
  --o-visualization alpha-rarefaction.qzv
  
# 稀释
qiime diversity core-metrics-phylogenetic \
  --i-phylogeny rooted-tree.qza \
  --i-table table.qza  \
  --p-sampling-depth 3000 \
  --m-metadata-file metadata.txt \
  --output-dir core-metrics
  
# ASV taxa assign
qiime feature-classifier classify-sklearn \
    --i-classifier ../ref_7/gg-13-8-99-nb-classifier.qza \
    --i-reads rep-seqs.qza \
    --o-classification taxonomy.qza
    
 qiime metadata tabulate \
    --m-input-file taxonomy.qza \
    --o-visualization taxonomy.qzv
    
 # collaspe taxa abundance
 qiime taxa barplot \
  --i-table core-metrics/rarefied_table.qza \
  --i-taxonomy taxonomy.qza \
  --m-metadata-file metadata.txt \
  --o-visualization rarefied_taxa-bar-plots.qzv
  
# ASV abundance
mkdir -p ASV; cd ASV
cp ../core-metrics/rarefied_table.qza rarefied_table.zip
unzip rarefied_table.zip;  # 需要对解压下的文件夹重命名，比如命名为rarefied_table
cd rarefied_table/data;
biom convert -i feature-table.biom -o asv_table.txt --to-tsv
```

## QIIME2分析

### *年龄和alpha多样性*

``` bash
qiime diversity alpha-correlation \
    --m-metadata-file metadata.txt \
    --i-alpha-diversity core-metrics/observed_otus_vector.qza \
    --o-visualization  core-metrics/observed_otus_correlation.qzv
```

![](https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/16S/Micor-develop/ref7_13/alpha-correlation.svg)

| Test               | Spearman |
| :----------------- | :------- |
| **Test statistic** | 0.6847   |
| **P-value**        | 0.0000   |
| **Sample size**    | 1161     |

整体而言，随着年龄的逐步提升，观测到的微生物多样性也是逐步上升的。

### 奇怪的PCoA图

> jaccard pcoa

![](https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/16S/Micor-develop/ref7_13/emperor-jaccard.svg)

> unweighted unifrac

![](https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/16S/Micor-develop/ref7_13/emperor-unweight-unifrac.svg)

上图中，红色是孟加拉数据，而橙色为其文中提到的验证马拉维数据。蓝色为马拉维数据。数据分成了两个接近平行的月牙，呈现出较为奇怪的特征。

## 随机森林模型

``` r
rm(list=ls())
library(randomForest)
library(ggplot2)
library(dplyr)

abundance <- read.table("asv_table.txt",header=T,sep="\t",check.names = F,row.names=1)
relative_abundance <- abundance/3000*100
count <- function(abun){
     num <- sum(abun>0.1)
     return(num)
}
filter_count <- apply(relative_abundance,1,count)
 
index <- which(filter_count>8) 
filter_abundance <- relative_abundance[index,]
metadata <- read.table("metadata.txt",header=T,sep="\t")
otu <- data.frame(t(filter_abundance))
otu$Age <- metadata[rownames(otu),]$Age

# train and valid samples
train_index <- sample(rownames(otu),nrow(otu)*0.7)
valid_index <- setdiff(rownames(otu),train_index) 
train_sample <- otu[train_index,]
valid_sample <- otu[valid_index,]

## 选择年龄区间为1-18月
#train_sample <- train_sample[which(train_sample$Age<18),]
#valid_sample <- valid_sample[which(valid_sample$Age<18),]

# RF Glimpse
set.seed(123)  # 创建随机种子，从而使得结果可重复
otu_train.forest<- randomForest(Age~., data = train_sample, importance = TRUE) ## 开始训练
otu_train.forest
```

>Call:
> randomForest(formula = Age ~ ., data = train_sample, importance = TRUE) 
>               Type of random forest: regression
>                     Number of trees: 500
>No. of variables tried at each split: 179
>
>          Mean of squared residuals: 14.92069
>                    % Var explained: 72.48

``` r
importance_otu <- data.frame(otu_train.forest$importance)
varImpPlot(otu_train.forest, n.var = min(30, nrow(otu_train.forest$importance)),
main = 'Top 30 - variable importance')
```

![](https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/16S/Micor-develop/ref7_13/variable_importance.png)

``` R
importance_otu <- importance_otu[order(importance_otu$IncNodePurity, decreasing = TRUE), ]
write.table(importance_otu, 'importance_otu.txt', sep = '\t', col.names = NA, quote = FALSE)

otu_train <- train_sample
otu_train.cv <- replicate(5, rfcv(otu_train[-ncol(otu_train)], otu_train$Age, cv.fold = 10, step = 1.5), simplify = FALSE)
otu_train.cv
otu_train.cv <- data.frame(sapply(otu_train.cv, '[[', 'error.cv'))
otu_train.cv$otus <- rownames(otu_train.cv)
otu_train.cv <- reshape2::melt(otu_train.cv, id = 'otus')
otu_train.cv$otus <- as.numeric(as.character(otu_train.cv$otus))
otu_train.cv.mean <- aggregate(otu_train.cv$value, by = list(otu_train.cv$otus), FUN = mean)
ggplot(otu_train.cv.mean, aes(Group.1, x)) +
	geom_line() +
	theme(panel.grid = element_blank(), panel.background = element_rect(color = 'black', fill = 'transparent')) +
	labs(title = '',x = 'Number of OTUs', y = 'Cross-validation error')
head(otu_train.cv.mean, 20)
```

>   Group.1        x  
>   1        1 45.04694  
>   2        2 35.18462  
>   3        3 28.71272  
>   4        4 25.08448  
>   5        6 21.38917  
>   6        9 18.89489  
>   7       14 17.99712  
>   8       21 17.23732  
>   9       31 16.97913  
>   10      47 16.47820  
>   11      71 15.81818  
>   12     106 15.55566  
>   13     159 15.45693  
>   14     239 15.39939  
>   15     358 15.45632  
>   16     537 15.55963   

21 is good.

### 21个变量的简约回归

``` R
importance_otu.select <- importance_otu[1:21, ]
otu_id.select <- rownames(importance_otu.select)
write.table(importance_otu.select, 'importance_otu.select.txt', sep = '\t', col.names = NA, quote = FALSE)
##只包含21个重要预测变量的简约回归
otu_train.select <- train_sample[,c(otu_id.select,"Age")]
otu_test.select <- valid_sample[,c(otu_id.select,"Age")]

#随机森林计算（默认生成 500 棵决策树），详情 ?randomForest
set.seed(123)
otu_train.select.forest <- randomForest(Age~., data = otu_train.select, importance = TRUE)
```

>otu_train.select.forest 

> Call:
>  randomForest(formula = Age ~ ., data = otu_train.select, importance = TRUE) 
>                Type of random forest: regression
>                      Number of trees: 500
> No. of variables tried at each split: 7
>
>           Mean of squared residuals: 16.86074
>                     % Var explained: 68.9

``` R
names <- rownames(otu_test.select)
age_predict <- predict(otu_train.select.forest, otu_test.select)
plot(otu_test.select$Age, age_predict, main = 'Test',xlab = 'age (Months)', ylab = 'Predict')
abline(1, 1)
```

![](https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/16S/Micor-develop/ref7_13/predict.png)

``` R
Test_data <- data.frame(otu_test.select$Age, age_predict)
Test_Fit <- lm(age_predict~otu_test.select.Age,data=Test_data)
summary(Test_Fit)
```

>Call:  
>lm(formula = age_predict ~ otu_test.select.Age, data = Test_data)    
>
>Residuals:  
>     Min       1Q   Median       3Q      Max   
>-10.4187  -1.9778   0.0438   1.8712   9.3173   
>
>Coefficients:  
>                    Estimate Std. Error t value Pr(>|t|)    
>(Intercept)          3.66695    0.29330   12.50   <2e-16 ***    
>
>otu_test.select.Age  0.71086    0.02236   31.79   <2e-16 ***    
>
>Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1   
>
>Residual standard error: 2.928 on 347 degrees of freedom  
>Multiple R-squared:  0.7444,	Adjusted R-squared:  0.7437   
>F-statistic:  1011 on 1 and 347 DF,  p-value: < 2.2e-16  

### 物种分析

``` R
ASV_taxa <- read.table("taxonomy.txt",header=T,sep="\t",check.names=F)
select <- read.table("importance_otu.select.txt",header=T,sep="\t",check.names=F)
result <- inner_join(ASV_taxa,select,by="ASV")
result <- result[order(result$IncNodePurity, decreasing = TRUE), ]
write.table(result,"Select_asv_taxa.txt",quote=F,sep="\t",row.names=F)
```

| ASV                              | Taxon                                                        |
| -------------------------------- | ------------------------------------------------------------ |
| f118662f3e5f209f582ffbaf39e21807 | Bacteria; Firmicutes; Clostridia; Clostridiales; Lachnospiraceae;  Blautia; s__ |
| c7dad10896f6c10619ec4d963d5a0d4c | Bacteria; Firmicutes; Clostridia; Clostridiales; Lachnospiraceae; Dorea;  s__longicatena |
| c3ae999022b2b3dba38bf8d4e08242c1 | Bacteria; Firmicutes; Clostridia; Clostridiales; Lachnospiraceae; Dorea;  s__formicigenerans |
| ed8930aed9ccc63f5b22fd4229ef3361 | Bacteria; Firmicutes; Clostridia; Clostridiales; Ruminococcaceae;  Gemmiger; s__formicilis |
| 6c38c23dde65be3184382ccaafdbc7aa | Bacteria; Firmicutes; Bacilli; Bacillales; Staphylococcaceae;  Staphylococcus |
| b99a5879168a9d117fdffb707e1f682c | Bacteria; Firmicutes; Clostridia; Clostridiales; Lachnospiraceae;  Blautia; s__ |
| 2792f7e5c84a91ba5a549be947465cb6 | Bacteria; Actinobacteria; Actinobacteria; Bifidobacteriales;  Bifidobacteriaceae; Bifidobacterium |
| f00219cfb5ea9d13f4be1e63c6822722 | Bacteria; Firmicutes; Clostridia; Clostridiales; Ruminococcaceae;  Faecalibacterium; s__prausnitzii |
| 7be3fed5631db826c499e1f0f3d5bff1 | Bacteria; Firmicutes; Clostridia; Clostridiales; Lachnospiraceae;  Roseburia; s__faecis |
| 2ea082fe7de3327a025e61006313746b | Bacteria; Proteobacteria; Gammaproteobacteria; Pasteurellales;  Pasteurellaceae; Haemophilus; s__parainfluenzae |
| c60a6281dea0f58a0154c2efcdb19cff | Bacteria; Firmicutes; Bacilli; Lactobacillales; Lactobacillaceae;  Lactobacillus; s__ruminis |
| c878ea3979de4d780d7e3e9b37797c9f | Bacteria; Actinobacteria; Actinobacteria; Bifidobacteriales;  Bifidobacteriaceae; Bifidobacterium; s__ |
| 98aaddfc3162f1cc9f308cc008748a28 | Bacteria; Firmicutes; Clostridia; Clostridiales; Lachnospiraceae;  Blautia; s__obeum |
| 3a9fc45622619d3fd38007c1852eb7cd | Bacteria; Bacteroidetes; Bacteroidia; Bacteroidales; Prevotellaceae;  Prevotella; s__copri |
| 28122e04ec3fd0344659c8d74bd7d78b | Bacteria; Firmicutes; Clostridia; Clostridiales; Lachnospiraceae |
| e96d365cd48141668838c3f8121fa21a | Bacteria; Firmicutes; Clostridia; Clostridiales; Lachnospiraceae;  Coprococcus; s__catus |
| fd0a3dc53155f8a85951b9ee8addb8b0 | Bacteria; Firmicutes; Bacilli; Lactobacillales; Lactobacillaceae;  Lactobacillus; s__mucosae |
| 1585491003a692e661e3e84794a0c541 | Bacteria; Proteobacteria; Gammaproteobacteria; Enterobacteriales;  Enterobacteriaceae |
| b97be820e4bf0d5a3c32cfb1f1512a40 | Bacteria                                                     |
| fb9fff85b91dae8de2154b82cd2ca4b5 | Bacteria; Firmicutes; Clostridia; Clostridiales; Ruminococcaceae;  Faecalibacterium; s__prausnitzii |
| 88bad5fd626727f0d8fef57f9dbb6e0f | Bacteria; Firmicutes; Bacilli; Lactobacillales; Enterococcaceae;  Enterococcus |

上述表格为按照`importance score`排列所得。

### Summary

多个数据集之间是可以合并分析的，并且具有较好的一致性。

### 数据子集分析（<18 months）

由于数据混合和时间跨度过大，猜测是否是年龄太大会带来拟合上的问题。故选取数据为18个月以内的数据进行分析

我们的变量存储在`otu`中,筛选后保留920个样本，包含745个孟加拉样本和175马拉维样本。

仍然以0.7和0.3的比例选取训练和测试样本。对应637和273个训练测试样本

``` R
otu_subset <- otu[which(otu$Age<=18),]
train_index <- sample(rownames(otu_subset),nrow(otu_subset)*0.7)
valid_index <- setdiff(rownames(otu_subset),train_index) 
train_sample <- otu_subset[train_index,]
valid_sample <- otu_subset[valid_index,]
# RF Glimpse
set.seed(123)  # 创建随机种子，从而使得结果可重复
otu_train.forest<- randomForest(Age~., data = train_sample, importance = TRUE) ## 开始训练
otu_train.forest
```

>Call:
> randomForest(formula = Age ~ ., data = train_sample, importance = TRUE) 
>               Type of random forest: regression
>                     Number of trees: 500
>No. of variables tried at each split: 179
>
>          Mean of squared residuals: 8.873464
>                    % Var explained: 64.2

``` R
importance_otu <- data.frame(otu_train.forest$importance)
importance_otu <- importance_otu[order(importance_otu$IncNodePurity, decreasing = TRUE), ]
otu_train <- train_sample
otu_train.cv <- replicate(5, rfcv(otu_train[-ncol(otu_train)], otu_train$Age, cv.fold = 10, step = 1.5), simplify = FALSE)
otu_train.cv
otu_train.cv <- data.frame(sapply(otu_train.cv, '[[', 'error.cv'))
otu_train.cv$otus <- rownames(otu_train.cv)
otu_train.cv <- reshape2::melt(otu_train.cv, id = 'otus')
otu_train.cv$otus <- as.numeric(as.character(otu_train.cv$otus))
otu_train.cv.mean <- aggregate(otu_train.cv$value, by = list(otu_train.cv$otus), FUN = mean)
ggplot(otu_train.cv.mean, aes(Group.1, x)) +
	geom_line() +
	theme(panel.grid = element_blank(), panel.background = element_rect(color = 'black', fill = 'transparent')) +
	labs(title = '',x = 'Number of OTUs', y = 'Cross-validation error')
head(otu_train.cv.mean, 20)
```

>      Group.1         x  
>   1        1 21.164391  
>   2        2 16.759359  
>   3        3 14.546684  
>   4        4 13.048069  
>   5        6 11.507501  
>   6        9 10.893205  
>   7       14 10.096992  
>   8       21  9.880045  
>   9       31  9.606862  
>   10      47  9.338943  
>   11      71  9.267957  
>   12     106  9.242369  
>   13     159  9.274685  
>   14     239  9.257842  
>   15     358  9.265815  
>   16     537  9.412580  

结果仍然是21.我们选择这21个进行回归测试：其回归变量解释为：

>Call:
> randomForest(formula = Age ~ ., data = otu_train.select, importance = TRUE) 
>         Type of random forest: regression
>               Number of trees: 500
>No. of variables tried at each split: 7

          Mean of squared residuals: 9.041955
                    % Var explained: 63.52
而测试集的结果为：

>Call:
>lm(formula = age_predict ~ otu_test.select.Age, data = Test_data)
>
>Residuals:
>    Min      1Q  Median      3Q     Max 
>-7.0720 -1.5880  0.0597  1.6154  6.9139 
>
>Coefficients:
>                    Estimate Std. Error t value Pr(>|t|)    
>(Intercept)          3.86618    0.28879   13.39   <2e-16 *****
>
>**otu_test.select.Age  0.52134    0.02876   18.13   <2e-16 ***
>
>Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
>
>Residual standard error: 2.417 on 271 degrees of freedom
>Multiple R-squared:  0.548,	Adjusted R-squared:  0.5464 
>F-statistic: 328.6 on 1 and 271 DF,  p-value: < 2.2e-16

在选取18月之前的数据作为分析发现，整体的回归性能变差了。

### Bangladesh only analysis

```R
otu_bangladesh <- na.omit(otu[rownames(metadata)[which(metadata$Type=="Bangladesh")],])
## 共895个样本
train_index <- sample(rownames(otu_bangladesh),nrow(otu_bangladesh)*0.7) #626 samples
valid_index <- setdiff(rownames(otu_bangladesh),train_index) # 269 samples
train_sample <- otu_bangladesh[train_index,]
valid_sample <- otu_bangladesh[valid_index,]
# RF Glimpse
set.seed(123)  # 创建随机种子，从而使得结果可重复
otu_train.forest<- randomForest(Age~., data = train_sample, importance = TRUE) ## 开始训练
otu_train.forest
```

>Call:
> randomForest(formula = Age ~ ., data = train_sample, importance = TRUE) 
>               Type of random forest: regression
>                     Number of trees: 500
>No. of variables tried at each split: 179
>
>          Mean of squared residuals: 12.52807
>                    % Var explained: 72.18

五折交叉验证的结构表明在以feature数目为9的时候就出现了明显的误差速率减少，为了便于比较，我们仍然选择21个feature进行计算。

>   Group.1        x  
>   1        1 34.08701  
>   2        2 25.09412  
>   3        3 20.06591  
>   4        4 18.30999  
>   5        6 16.55228  
>   6        9 15.56560  
>   7       14 14.50420  
>   8       21 14.25866  
>   9       31 14.07609  
>   10      47 13.83201  
>   11      71 13.42821  
>   12     106 13.42704  
>   13     159 13.34986  
>   14     239 13.36199  
>   15     358 13.48967  
>   16     537 13.49233  

上述结果为：

>Call:
> randomForest(formula = Age ~ ., data = otu_train.select, importance = TRUE) 
>               Type of random forest: regression
>                     Number of trees: 500
>No. of variables tried at each split: 7
>
>          Mean of squared residuals: 13.46171
>                    % Var explained: 70.11

该结果是较为精准的。

#### 结果比较

我们将仅由Bangladesh的数据和合并数据的物种ASV进行比较，其中14个物种是共有的。

| Taxon_Different                  | Confidence                                                   |
| -------------------------------- | ------------------------------------------------------------ |
| ca033ba99529a006cd0c11e92fb8e942 | Bacteria; Firmicutes; Erysipelotrichi; Erysipelotrichales;  Erysipelotrichaceae; Catenibacterium; s__ |
| b6a27409a4e20ef37eec8b1c67c6c076 | Bacteria                                                     |
| 648ca9256127f35fad4013e8f62cef3b | Bacteria                                                     |
| 1c4b3a88cb2d430b45d61279400da346 | Bacteria                                                     |
| 1b369a94dfbe2959f70aa4e147b09420 | Bacteria; Firmicutes; Bacilli; Lactobacillales; Streptococcaceae;  Streptococcus; s__luteciae |
| dea94664b6e8ad6823604fa2c8f2eb4e | Bacteria; Firmicutes; Bacilli; Lactobacillales; Streptococcaceae;  Streptococcus; s__ |
| 5d5e07a2af7eac6b0b0602d5a163a02e | Bacteria; Actinobacteria; Actinobacteria; Bifidobacteriales;  Bifidobacteriaceae; Bifidobacterium; s__bifidum |

| Taxon_Same                       | Confidence                                                   |
| -------------------------------- | ------------------------------------------------------------ |
| f118662f3e5f209f582ffbaf39e21807 | Bacteria; Firmicutes; Clostridia; Clostridiales; Lachnospiraceae;  Blautia; s__ |
| c3ae999022b2b3dba38bf8d4e08242c1 | **Bacteria; Firmicutes; Clostridia; Clostridiales; Lachnospiraceae; Dorea;  s__formicigenerans** |
| c7dad10896f6c10619ec4d963d5a0d4c | **Bacteria; Firmicutes; Clostridia; Clostridiales; Lachnospiraceae; Dorea;  s__longicatena** |
| 6c38c23dde65be3184382ccaafdbc7aa | **Bacteria; Firmicutes; Bacilli; Bacillales; Staphylococcaceae;  Staphylococcus** |
| f00219cfb5ea9d13f4be1e63c6822722 | **Bacteria; Firmicutes; Clostridia; Clostridiales; Ruminococcaceae;  Faecalibacterium; s__prausnitzii** |
| c60a6281dea0f58a0154c2efcdb19cff | **Bacteria; Firmicutes; Bacilli; Lactobacillales; Lactobacillaceae;  Lactobacillus; s__ruminis** |
| c878ea3979de4d780d7e3e9b37797c9f | **Bacteria; Actinobacteria; Actinobacteria; Bifidobacteriales;  Bifidobacteriaceae; Bifidobacterium; s__** |
| b99a5879168a9d117fdffb707e1f682c | Bacteria; Firmicutes; Clostridia; Clostridiales; Lachnospiraceae;  Blautia; s__ |
| 2ea082fe7de3327a025e61006313746b | **Bacteria; Proteobacteria; Gammaproteobacteria; Pasteurellales;  Pasteurellaceae; Haemophilus; s__parainfluenzae** |
| ed8930aed9ccc63f5b22fd4229ef3361 | Bacteria; Firmicutes; Clostridia; Clostridiales; Ruminococcaceae;  Gemmiger; s__formicilis |
| b97be820e4bf0d5a3c32cfb1f1512a40 | Bacteria                                                     |
| 2792f7e5c84a91ba5a549be947465cb6 | **Bacteria; Actinobacteria; Actinobacteria; Bifidobacteriales;  Bifidobacteriaceae; Bifidobacterium** |
| fd0a3dc53155f8a85951b9ee8addb8b0 | **Bacteria; Firmicutes; Bacilli; Lactobacillales; Lactobacillaceae;  Lactobacillus; s__mucosae** |
| 3a9fc45622619d3fd38007c1852eb7cd | Bacteria; Bacteroidetes; Bacteroidia; Bacteroidales; Prevotellaceae;  Prevotella; s__copri |

上述14个保留物种在原文中top24的以加粗形式显示。

## 分阶段随机森林分类

上述模型是以每个微生物（细菌）的丰度和年龄进行的回归，这里我们选择分阶段做分类任务预测。

整个数据的年龄分布:



<img src="https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/16S/Micor-develop/ref7_13/classification/age_distribution.png" style="width:30px height:20px" />

载入packages：

``` R
library(randomForest)
require(caTools)
library(ROSE) ## 解决样本不均衡问题
library(doBy)  #方便基于分组取均值，使用其中的summaryBy()命令
library(reshape2)    #方便排列表格，使用其中的melt()命令
library(ggplot2)
library(pROC)  # 绘制ROC曲线
library(caret)
library(pheatmap)
```



### 随机森林构建

创建stage:

``` R
metadata$stage <- NA
for(i in 1:dim(metadata)[1])
{
    if(metadata$Age[i]<=1) metadata$stage[i] <- "I"
    if(metadata$Age[i]>1 & metadata$Age[i]<=3) metadata$stage[i] <- "II"
    if(metadata$Age[i]>3 & metadata$Age[i]<=6) metadata$stage[i] <- "III"
    if(metadata$Age[i]>6 & metadata$Age[i]<=12) metadata$stage[i] <- "IV"
    if(metadata$Age[i]>12 & metadata$Age[i]<=18) metadata$stage[i] <- "V"
}
for(i in 1:dim(metadata)[1])
{
    if(metadata$Age[i]<=3) metadata$stage[i] <- "I"
    if(metadata$Age[i]>3 & metadata$Age[i]<=6) metadata$stage[i] <- "II"
    if(metadata$Age[i]>6 & metadata$Age[i]<=12) metadata$stage[i] <- "III"
    if(metadata$Age[i]>12 & metadata$Age[i]<=18) metadata$stage[i] <- "IV"
    if(metadata$Age[i]>18) metadata$stage[i] <- "V"
}
metadata <- na.omit(metadata)
otu <- data.frame(t(filter_abundance))
otu$stage <- metadata[rownames(otu),]$stage
otu <- na.omit(otu) # 保留了910个样本
table(otu$stage)
```

| I    | II   | III  | IV   | V    |
| ---- | ---- | ---- | ---- | ---- |
| 58   | 110  | 156  | 340  | 246  |

``` r
set.seed(1234)  ##确保结果可重复
train_index <- sample(rownames(otu),nrow(otu)*0.7)
valid_index <- setdiff(rownames(otu),train_index) 
train_sample <- transform(otu[train_index,],stage=as.factor(stage))
valid_sample <- transform(otu[valid_index,],stage=as.factor(stage))

rf_ntree <- randomForest(stage~.,data=train_sample,ntree=2000)
plot(rf_ntree)
```

![](https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/16S/Micor-develop/ref7_13/classification/error_rate.png)

当ntree=1000时，模型内的误差还不够稳定，另外，整体错误率十分的高，具有非常差的分类效果。我们勉强采取`ntree=500`进行计算

``` R
develop_rf <- randomForest(stage ~ ., train_sample, ntree=500,proximity=TRUE)
# develop_rf <- randomForest(stage ~ ., train_sample, ntree=1000,proximity=TRUE)
develop_rf
```

>Call:
> randomForest(formula = stage ~ ., data = train_sample, ntree = 500,      proximity = TRUE) 
>         Type of random forest: classification
>               Number of trees: 500
>No. of variables tried at each split: 23
>
>​    OOB estimate of  error rate: 40.97%
>
>Confusion matrix:
>     I II III  IV   V class.error
>I   21  7   2   7   2   0.4615385
>II   4 18  27  20   1   0.7428571
>III  0 20  40  42   1   0.6116505
>IV   3  9  15 184  34   0.2489796
>V    0  0   1  66 113   0.3722222

上述展示的OOB error为40.97%，有点惨不忍睹。这可能是由于多分类任务的复杂性带来的。

``` R
# 计算模型变量的重要性
importance_otu <- data.frame(importance(develop_rf))
importance_otu$accu <- NA
importance_otu <- importance_otu[order(importance_otu$MeanDecreaseGini, decreasing = T), ]

# 模型自身测试
pred1 <- predict(develop_rf, data = train_sample)
freq1 <- table(pred1, train_sample$stage)
sum(diag(freq1)/sum(freq1))
# the result is 0.5902669

test_predict <- predict(develop_rf, newdata = valid_sample)
cm = table(valid_sample$stage, test_predict)
```

|         | I    | II   | III  | IV   | V    |
| ------- | ---- | ---- | ---- | ---- | ---- |
| **I**   | 16   | 3    | 0    | 0    | 0    |
| **II**  | 9    | 15   | 8    | 7    | 1    |
| **III** | 0    | 7    | 21   | 22   | 3    |
| **IV**  | 2    | 1    | 5    | 77   | 10   |
| **V**   | 0    | 1    | 2    | 26   | 37   |

上述混淆矩阵的特色为，单个分类可能会错分到临近的其他分类，但是`I`很难分到`V`,即时间跨度大的错误分类很少出现。

这引出了一个猜想为，在婴儿的微生物发展过程中，可能并未有明显的性别特征能够将数据划分为不同的阶段，而是处于一种连续的渐变过程，但是总体上，如果两个阶段距离较远，则很可能具有较大的微生物群落差异。

### 变量筛选

借助于连续性的回归模型，我们还是选择使用部分ASV来查看这些样本的变化。我们选择50个ASV来进行随机森林构建，查看结果。

``` R
importance_otu_50 <- importance_otu[1:50, ]
otu_select <- rownames(importance_otu_50)
otu_train_50 <- train_sample[c(otu_select, 'stage')]
otu_valid_50 <- valid_sample[c(otu_select, 'stage')]
otu_train.forest_50 <- randomForest(stage ~ ., data = otu_train_50, importance = TRUE,proximity=TRUE)
valid_predict_50 <- predict(otu_train.forest_50, newdata = otu_valid_50)
cm = table(otu_valid_50$stage, valid_predict_50) # 混淆矩阵
sum(diag(cm)/sum(cm))  ## 0.5494505
cm
```

|         | I    | II   | III  | IV   | V    |
| ------- | ---- | ---- | ---- | ---- | ---- |
| **I**   | 16   | 3    | 0    | 0    | 0    |
| **II**  | 13   | 10   | 8    | 8    | 1    |
| **III** | 1    | 6    | 21   | 22   | 3    |
| **IV**  | 3    | 1    | 10   | 67   | 14   |
| **V**   | 0    | 1    | 0    | 29   | 36   |

我们想看一下这个物种都属于什么：

``` R
write.table(importance_otu_50,"classification/importance_50_otu.txt",sep="\t",quote=F)
ASV_taxa <- read.table("taxonomy.txt",header=T,sep="\t",check.names=F)
select <- read.table("classification/importance_50_otu.txt",header=T,sep="\t",check.names=F)
result <- inner_join(ASV_taxa,select,by="ASV")
result <- result[order(result$MeanDecreaseGini, decreasing = TRUE), ]
write.table(result,"classification/Select_asv_taxa.txt",quote=F,sep="\t",row.names=F)
```

物种结果可见于[文本](https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/16S/Micor-develop/ref7_13/classification/Select_asv_taxa.txt)

